flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java

```diff
@@ -0,0 +1,146 @@
+import org.apache.flink.runtime.executiongraph.failover.flip1.FailureHandlingResult;
+import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nullable;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * {@code FailureHandlingResultSnapshot} creates a snapshot of a {@link FailureHandlingResult}
+ * providing the actual {@link Execution Executions}.
+ */
+public class FailureHandlingResultSnapshot {
+
+    private final ExecutionAttemptID rootCauseExecution;
+    private final Throwable rootCause;
+    private final long timestamp;
+    private final Map<ExecutionAttemptID, Execution> failedExecutions;
+
+    /**
+     * Creates a {@code FailureHandlingResultSnapshot} based on the passed {@link
+     * FailureHandlingResult} and {@link ExecutionVertex ExecutionVertices}.
+     *
+     * @param failureHandlingResult The {@code FailureHandlingResult} that is used for extracting
+     *     the failure information.
+     * @param executionVertices The {@code ExecutionVertices} that shall be scanned for failures
+     *     based on the passed {@code FailureHandlingResult}.
+     * @return The {@code FailureHandlingResultSnapshot}.
+     */
+    public static FailureHandlingResultSnapshot create(
+            FailureHandlingResult failureHandlingResult,
+            Iterable<ExecutionVertex> executionVertices) {
+        final ExecutionVertexID rootCauseExecutionVertexId =
+                failureHandlingResult.getExecutionVertexIdOfFailedTask().orElse(null);
+
+        ExecutionAttemptID rootCauseExecutionAttemptId = null;
+        final Map<ExecutionAttemptID, Execution> failedExecutions = new HashMap<>();
+        for (ExecutionVertex executionVertex : executionVertices) {
+            if (executionVertex.getID().equals(rootCauseExecutionVertexId)) {
+                Preconditions.checkArgument(
+                        executionVertex.getCurrentExecutionAttempt().getFailureInfo().isPresent(),
+                        String.format(
+                                "The execution %s didn't provide a failure info even though the corresponding ExecutionVertex %s is marked as having handled the root cause of this failure.",
+                                executionVertex.getCurrentExecutionAttempt().getAttemptId(),
+                                executionVertex.getID()));
+                rootCauseExecutionAttemptId =
+                        executionVertex.getCurrentExecutionAttempt().getAttemptId();
+            }
+
+            final Execution execution = executionVertex.getCurrentExecutionAttempt();
+            if (failureHandlingResult.getVerticesToRestart().contains(executionVertex.getID())
+                    && execution.getFailureInfo().isPresent()) {
+                failedExecutions.put(execution.getAttemptId(), execution);
+            }
+        }
```

Comment:
I don't think that this is the most efficient way to look up the required information. If the EG is quite large and the failure region is small, then it will be quite costly. Maybe instead of passing in an `Iterable<ExecutionVertex>` we could pass in a lookup function `Function<ExecutionVertexID, ExecutionVertex>` which uses the `JobVertexID` and the `subtaskIndex` to look up the `ExecutionVertex`. This is not super important but if this is easy to change, it might save us some CPU cycles.